holmes/plugins/toolsets/rabbitmq/api.py

```diff
@@ -0,0 +1,411 @@
+
+    # For internal use
+    connection_status: Optional[ClusterConnectionStatus] = None
+    connection_error: Optional[str] = None
+
+
+class Partition(BaseModel):
+    node: str
+    unreachable_nodes: List[str]  # Nodes that 'node' cannot reach
+
+
+class NodeStatus(BaseModel):
+    node: str
+    running: bool  # Status as reported by the primary connected node
+
+
+class NodeInfo(BaseModel):
+    name: str
+    type: str
+    running: bool
+    mem_used: Optional[int] = None
+    mem_limit: Optional[int] = None
+    mem_alarm: bool
+    disk_free: Optional[int] = None
+    disk_free_limit: Optional[int] = None
+    disk_free_alarm: bool
+    fd_used: Optional[int] = None
+    fd_total: Optional[int] = None
+    sockets_used: Optional[int] = None
+    sockets_total: Optional[int] = None
+    uptime: Optional[int] = None
+    partitions: Optional[List[Any]] = None
+    error: Optional[str] = None
+
+
+class ClusterStatus(BaseModel):
+    nodes: List[NodeStatus]  # Overall node running status from primary view
+    network_partitions_detected: bool = False
+    partition_details: List[Partition]  # Combined list of detected partitions
+    raw_node_data: List[NodeInfo]  # Data from the primary connected node
+
+
+# --- Helper Functions (Slight modifications) ---
+
+
+def get_auth(config: RabbitMQClusterConfig) -> Optional[HTTPBasicAuth]:
+    if config.username or config.password:
+        return HTTPBasicAuth(
+            config.username or "guest",
+            config.password or "guest",
+        )
+    else:
+        return None
+
+
+def get_url(base_url: str, endpoint: str) -> str:
+    """Constructs a URL using a base and an endpoint."""
+    # Ensure base_url ends with '/' for urljoin to work predictably
+    if not base_url.endswith("/"):
+        base_url += "/"
+    return urljoin(base_url, endpoint.lstrip("/"))
+
+
+@backoff.on_exception(
+    backoff.expo,
+    requests.exceptions.RequestException,
+    max_tries=3,
+    giveup=lambda e: isinstance(e, requests.exceptions.HTTPError)
+    and e.response.status_code < 500,
+)
+def make_request(
+    config: RabbitMQClusterConfig,
+    method: str,
+    url: str,
+    params: Optional[Dict] = None,
+    data: Optional[Dict] = None,
+) -> Any:
+    """Makes an HTTP request to the RabbitMQ Management API."""
+    headers = {"Content-Type": "application/json"}
+    try:
+        response = requests.request(
+            method=method,
+            url=url,
+            headers=headers,
+            auth=get_auth(config),
+            params=params,
+            json=data,
+            timeout=config.request_timeout_seconds,
+            verify=config.verify_certs,
+        )
+        response.raise_for_status()
+        return response.json()
+    except requests.exceptions.RequestException as e:
+        logging.error(f"Request failed for {method} {url}: {e}")
+        raise  # Re-raise after logging for upstream handling
+
+
+def node_data_to_node_info(node_data: Dict) -> NodeInfo:
```

Comment:
I think pydantic can do that kind of logic. not entirely sure. not blocking the PR but worth checking

Reply:
Not with the current `NodeInfo` struct but you make a good point on that code. I modified the struct so that it can directly receive node_data with `NodeInfo(**node_data)`.